ALGORITHM.md

```diff
@@ -0,0 +1,151 @@
+# The FSRS Scheduler Algorithm
+
+## Overview
+
+The FSRS scheduler algorithm is a Markov chain with several state variables and four different transition events. It is based on a variant of the **DSR** (Difficulty, Stability, Retrievability) model which is used to predict human memory states.
+
+![The FSRS Scheduler Algorithm State Diagram](fsrs-diagram.png)
+
+## The Algorithm
+
+### <ins>Variables</ins>
+
+$DECAY=-0.5,$
+
+$FACTOR = 0.9^{\frac{1}{DECAY}}-1$,
+
+$G$: Grade (card rating)
+- $1$: again
+- $2$: hard
+- $3$: good
+- $4$: easy
+
+State:
+- New
+- Learning
+- Review
+- Relearning
+
+Scheduler weights: $(w_0, w_1,... w_{18})$
+
+`HARD PENALTY` = $w_{15}$
+
+`EASY BONUS` = $w_{16}$
+
+$S$: Stability
+
+$D$: Difficulty
+
+$R$: Retrievability (probability of recall)
+
+$r$: Request retention
+
+$t$: Days since last review ("elapsed days")
+
+$I$: Interval ("scheduled days"). Number of days before the card is due next
+
+
+### <ins>Functions</ins>
+
+<ins>Initial difficulty</ins>: 
+
+$D_0(G) = w_4-e^{w_5\cdot(G-1)}+1$
+
+<ins>Initial stability</ins>: 
+
+$S_0(G) = w_{G-1}$
+
+<ins>Next interval</ins>: 
+
+$I(r,S) = \frac{S}{FACTOR}\cdot\left(r^{\frac{1}{DECAY}}-1\right)$
+
+<ins>Mean reversion</ins>: 
+
+$MR(a,b) = w_7\cdot a + (1-w_7)\cdot b$
+
+<ins>Short-term stability</ins>: 
+
+$S^\prime(S,G) = S\cdot e^{w_{17}\cdot (G-3+w_{18})}$
+
+<ins>Current retrievability</ins>: 
+
+$R(t,S) = \left(1+FACTOR\cdot \frac{t}{S}\right)^{DECAY}$
+
+<ins>Forget stability</ins>: 
+
+$S^\prime_f(D,S,R) = w_{11} \cdot D^{-12}\cdot \Big[(S+1)^{w_{13}} -1 \Big]\cdot e^{w_{14}\cdot (1-R)}$
+
+<ins>Recall stability</ins>: 
+
+$$S^\prime_r(D,S,R,G) = S\cdot \left[1+e^{w_{8}}\cdot (11-D)\cdot S^{-w_9}\cdot (e^{w_{10}\cdot (1-R)}-1)\cdot \textrm{HARD PENALTY(if $G$=2)}\cdot \textrm{EASY BONUS(if $G$=4)} \right]$$
+
+---
+### State transitions
+
+**<ins>New card rated Easy:</ins>**
+
+State <- Review
+
+$S=S_0(4)$
+
+$D=D_0(4)$
+
+$I = I(r,S)$
+
+**<ins>New card rated Again / Hard / Good</ins>:**
+
+State <- Learning
+
+$S = S_0(G)$
+
+$D = D_0(G)$
+
+$I = 0$
+
+**<ins>Learning / Relearning card rated Again / Hard:</ins>**
+
+State <- Learning / Relearning
+
+$S = S^\prime(S,G)$
+
+$D = MR\Big(D_0(4), D-w_6\cdot (G-3)\Big)$
```

Comment:
The `repeat` method does not directly call `mean_reversion`. I suggest changing it to use `next_difficulty`, as follows:  
```suggestion
$D = D^\prime(D,G)$
```